chain-impl-mockchain/src/block.rs

```diff
@@ -277,19 +286,91 @@ impl property::Deserialize for SignedBlockSummary {
 }
 
 impl property::HasTransaction for Block {
-    type Transactions = [SignedTransaction];
-    fn transactions(&self) -> &Self::Transactions {
-        &self.transactions
+    type Transaction = SignedTransaction;
+    fn transactions<'a>(&'a self) -> Box<Iterator<Item = &SignedTransaction> + 'a> {
+        Box::new(self.contents.iter().filter_map(|msg| match msg {
+            Message::Transaction(tx) => Some(tx),
+            _ => None,
+        }))
     }
 }
 
 impl property::HasTransaction for SignedBlock {
-    type Transactions = [SignedTransaction];
-    fn transactions(&self) -> &Self::Transactions {
+    type Transaction = SignedTransaction;
+    fn transactions<'a>(&'a self) -> Box<Iterator<Item = &SignedTransaction> + 'a> {
         self.block.transactions()
     }
 }
 
+pub const TAG_TRANSACTION: u8 = 1;
+pub const TAG_STAKE_KEY_REGISTRATION: u8 = 2;
+pub const TAG_STAKE_KEY_DEREGISTRATION: u8 = 3;
+pub const TAG_STAKE_DELEGATION: u8 = 4;
+pub const TAG_STAKE_POOL_REGISTRATION: u8 = 5;
+pub const TAG_STAKE_POOL_RETIREMENT: u8 = 6;
+
+impl property::Serialize for Message {
+    type Error = std::io::Error;
+    fn serialize<W: std::io::Write>(&self, writer: W) -> Result<(), Self::Error> {
+        use chain_core::packer::*;
+        let mut codec = Codec::from(writer);
+        match self {
+            Message::Transaction(signed) => {
+                codec.put_u8(TAG_TRANSACTION)?;
```

Comment:
Are tagged values better here, rather than serializing events of different types in separate (possibly type-tagged) arrays?